okio/src/main/java/okio/Okio.java

```diff
@@ -242,4 +245,50 @@ static boolean isAndroidGetsocknameError(AssertionError e) {
     return e.getCause() != null && e.getMessage() != null
         && e.getMessage().contains("getsockname failed");
   }
+
+  public static Source tree(final Source source, final BufferedSink cacheBody)
+  {
+    return new Source() {
+      private boolean cacheRequestClosed;
+
+      @Override
+      public long read(Buffer sink, long byteCount) throws IOException {
+        long bytesRead;
+        try {
+          bytesRead = source.read(sink, byteCount);
+        } catch (IOException e) {
+          if (!cacheRequestClosed) {
+            cacheRequestClosed = true;
+          }
+          throw e;
+        }
+
+        if (bytesRead == -1) {
+          if (!cacheRequestClosed) {
+            cacheRequestClosed = true;
+            cacheBody.close(); // The cache response is complete!
+          }
+          return -1;
+        }
+
+        sink.copyTo(cacheBody.buffer(), sink.size() - bytesRead, bytesRead);
+        cacheBody.emitCompleteSegments();
+        return bytesRead;
+      }
+
+      @Override
+      public Timeout timeout() {
+        return source.timeout();
+      }
+
+      @Override
+      public void close() throws IOException {
+        if (!cacheRequestClosed
+                && !discard(this, DISCARD_STREAM_TIMEOUT_MILLIS, MILLISECONDS)) {
```

Comment:
I'm not sure this should be part of the feature itself. If you fail to exhaust the stream then I don't think it's Okio's job to try and continue it for you. The 100ms policy is of OkHttp requests, and doesn't make sense everywhere.